### DISCUSSION TITLE
The future of the tree adapter architecture

### DISCUSSION THREAD

**Discussion by 43081j** [parse5/parse5-fork#38] [General]
Currently, tree adapters exist as a concept so we can instruct the parser on how to build up an AST, any AST.

This allows us to have parse5 produce its own AST or a htmlparser2 AST for example.

Sounds like an outcome we may want, but im feeling more and more like its definitely the wrong solution to the problem.

## It isn't possible to strongly type it

Our current types, and the ones in DT (which i wrote) are _wrong_.

All along, we thought this pattern worked:

```ts
interface TypeMap {
  element: unknown;
  comment: unknown;
}

interface Adapter<T extends TypeMap> {
  getAttribute(el: T['element'], name: string): string;
}
```

but it doesn't. all the types narrow to `unknown`, the compiler will allow you to `getAttribute(commentNode, 'foo')` because `T['comment']` and `T['element']` both ultimately resolve to `unknown`.

This isn't fixable.

Type maps won't work, tuples of types won't work.

I talked to a few people on typescript discord and we came up with this:

```ts
interface Adapter<
  TDocument,
  TElement,
  TComment,
  ...
> {
  getAttribute(el: TElement, name: string): string;
  isElement(node: TNode): node is TElement;
}
```

but this doesn't work either. because `TNode` can't be defined. where/how can we define it? it can't be a type parameter, because then it could be _anything_ (yet again) which means `TElement` may not necessarily be assignable to it, so tsc will fail.

we would have to repeat ourselves with the full union every time we reference `TNode`, and no longer have a `TNode`. craziness.

## It increases complexity, bugs, etc

JS parsers generally implement an esprima-compatible AST, because having a bunch of unrelated, incompatible ASTs would be a terrible idea. They did the right thing.

By not having one standard AST like this, we have a crazy amount of over-genericised complexity trying to handle the fact that everything can be anything. This means we definitely have more bugs, more complexity and lower performance than if we had a well known data structure.

## Consumers need to read docs in at least two places

A consumer needs to read the docs for the parser _and_ the tree adapter to find out what will happen.

# Solution

We should remove tree adapters entirely, making parse5 produce a single, standard AST.

We can then introduce the concept of an 'output transform' which can consume the parse5 AST and output _something else_.

This would lead to types like so:

```ts
import {htmlparser2Transform} from 'wherever';

const doc = parse(source: string): Document; // so simple

htmlparser2Transform; // (doc: Document) => T

const htmlparser2Doc = htmlparser2Transform(doc); // T
```

meaning we don't care anymore about all this over-generic nonsense until the very end, and even then its just a simple mapper function.

cc @fb55

**Comment by 43081j** [comment]
A friend did point out that this would be less efficient. If we do stick with the adapter pattern, I'm happy with that but we really need to figure the strong typing out. It is currently broken

**Reply by fb55** [reply]
I agree that removing adapters will likely make things slower, plus increase memory usage. I have a preference to to keep them around.

**Reply by 43081j** [reply]
alternatives attempted so far:

### typed adapter (`Adapter<TDocument, TElement, TComment, ...>`)

works, i think, but causes hell for us as every function in the parser has to list all these individual type parameters. every place that has a `Parser` needs to do the same, and every place that has an adapter, etc, etc.

**Reply by 43081j** [reply]
@wooorm @fb55 i know i have shot down almost every comment so far but thats just because this is a difficult problem. im happy with us keeping adapters, _if_ we can fix the types.

i've spent a lot of time already trying to figure this out, with various people too. there is no way we can make "type maps" work, they were wrong from the start (my fault, i wrote them before this fork existed). 

if we go with passing every type parameter, it does work but we lose `Node`, which is unacceptable.

so i don't have any suggestions left yet, this is a tough problem..

**Reply by 43081j** [reply]
To summarise, rather than the several on going threads at once:

### `Adapter<TDoc, TEl, TComm, ...>`

Will work, but means everything has to have all of these type parameters. Including:

* `Parser<TDoc, TEl, TComm, ...>`
* `Serializer<TDoc, TEl, TComm, ...>`
* `ParserOptions<TDoc, TEl, TComm, ...>`
* All internal parser functions
* etc

This solution also means we lose `Node`. That is unacceptable but there may be a middle ground:

* Lose `Node` on tree adapters (specify the full union every time)
* But still expose it in each tree adapter's types (i.e. `Document` would still contain `Node[]`, a type the tree adapter exports)

So this solution may work at the expense of us writing and repeating an a LOT of parameters.

### Type maps

They won't work.

One way or another, the map has to have a type, which means `any` or `unknown` (since we can't expect any kind of base type all adapters implement). Due to this, we lose most of our type strength because everything becomes the base type (`any`/`unknown`), so most methods begin accepting anything rather than being strongly typed.

I'm not aware of a way to ever make this work.

### Dumb ParserOptions['treeAdapter'] types

Another option is to massively dumb down the tree adapter of the parser options.

This means each tree adapter would be strongly typed, _but would not implement any kind of common interface_.

The tree adapter property in our options object would then be "whatever" (impossible to strongly type at that point really:

```ts
parse<T>(
  code: string,
  options: {
    adapter: T
  }
): T extends {createDocument(): infer TDocument} ? TDocument : unknown;
```

basically [this](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwAdYBnEAHgBUA+ACgCh54wdQAueYjGLVAcwBoG8HAWx5i7AN5DGUYFFEgY7CkIC+dAJQr4IAB4YQqYMXiSwcKIYAiOMMgC2RjDW3weiJfAq37T1Bhq8AD83r6OzvDsaADWqDgA7qgA3HR0PIYwiFBgCADCyJw4DgCC8oowZkIWIFYg4f4ubpIYAJ4EIOwA5MB2XWqpjLwgGCUYXFgARsiGNPFsZm0d3SAQII39-PCoUE7snNx8bgc8vKkadMyonPBwxMgQGPAAvIQkIDRdXVuScgqZUiCUFMBSKpXKmTUmlSdDuDwwADolghnqj4D0+skgA)

@wooorm this one is my suggestion, i think. this way we can keep things simple and all we lose is some internal typing.

**Comment by wooorm** [comment]
I don’t really understand what would be impossible in TypeScript?
And if the only thing is `isElement(node: TNode)` vs `isElement(node: unknown)`, that sounds rather small to me? Not a good reason to change everything?

---

I’m assuming that removing adapters will cut down on bundle size and likely will improve performance as well: *just* building an object model instead of calling functions for everything should be faster (although this needs to be verified).
I’d recommend not changing too many things in one major version, though.

**Reply by 43081j** [reply]
that's not the only thing.

what isn't really possible (cleanly at least) is having this whole 'type map' concept work strongly.

in reality, what we need is this:

```ts
interface Adapter<TDocument, TElement, TComment, TDocumentFragment, TDocumentType, TTextNode, ...> {
```

but at some point, we decided to go with this instead:

```ts
interface TypeMap {
  document: unknown;
  element: unknown;
  comment: unknown;
  ...
}

interface Adapter<T extends TypeMap> {
```

this simply doesn't work. because they're all ultimately unknown. there's no type you can give them to make this work. 

it is what we have now, but is wrong and is actually hiding a bunch of type errors. i changed locally to have a strong fixed type and found a lot of type errors scattered all over the place. 

why don't we just have a dozen type params and let the compiler infer them, you might ask. we'd be in a better position than we are now... but we're then left with this problem:

```ts
interface Adapter<TDoc, TEl, TCom, ...> {
  appendChild(node: Node): void; // This can't work, because what is `Node`?
}

// ok lets pass that too you might say:

interface Adapter<TDoc, TEl, TCom, TNode, ...> {
  isElement(node: TNode): node is TEl; // This can't work! because `TNode` could be anything, `TEl` isn't assignable to it
}
```

so what we're left with is we can't have a `Node` type, we have to re-specify the union everywhere (`Document|Element|Comment|...`).

basically, we've never had working types for adapters. we just thought we did. so we either need to solve that or not have adapters really... otherwise we're hiding and causing all sorts of compiler errors/problems

**Reply by wooorm** [reply]
> this simply doesn't work. 

What doesn’t work? I feel that `unknown` matches the reality for parse5, it *should* treat nodes as unknown as adapters could generate whatever they want?

> and is actually hiding a bunch of type errors

Can you expand on what those are?

> we have to re-specify the union everywhere 

What’s bad about that? You can also classify certain unions as their own types, e.g., `Content` / `Parent` / `Child`?

> so we either need to solve that or not have adapters really.

I understand that it’s currently broken, but there are two options (a: fix them, b: remove adapters), and I don’t understand what sways you towards option b?

**Reply by fb55** [reply]
I wouldn't want to remove adapters, as there is too much other work based on them. Eg. JSDOM uses adapters extensively, with hooks for scripts. Getting the same behavior with the default adapter seems quite cumbersome.

> > and is actually hiding a bunch of type errors
>
> Can you expand on what those are?

Most issues seem to be in the open element stack, where we return a `parentNode`, with the type `element`.

**Reply by 43081j** [reply]
@wooorm it doesn't work because `T['element']` and `T['document']` etc become the same type. meaning they become meaningless, we may as well type everything as consuming and outputting `unknown` in that case. that really shouldn't be an option here, essentially giving consumers untyped AST.

what's bad about re-specifying the `Node` union rather than having a `Node` type alias for it? the fact that countless parse5 consumers and parse5 itself depends heavily on the `Node` type. if there is no longer one, we and everyone else has to know to specify the whole union every time.

what sways me to option b? nothing, i've mentioned im happy to do either.

@fb55 the open element stack is a container of elements _and_ document (one document, initially). this means all its types really consume and output `TElement|TDocument`. this changes a whole bunch of code when typed properly, meaning we have to cast or type guard in many places. anyway thats only one of the currently wrong types.

**Reply by wooorm** [reply]
`T['element']` should indeed internally be `unknown`? If a (typed) adapter is used, then the user would get a typed output?

> The fact that countless parse5 consumers and parse5 itself depends heavily on the Node type.

But, as the types are currently broken, and we’re looking at a major release and fixing them, that’d be an acceptable trade off?

> meaning we have to cast or type guard in many places. anyway thats only one of the currently wrong types.

A tuple can be used for that right?

```ts
type Stack = [1, ...2[]]

function x(list: Stack): void {
    const head = list[0] // 1
    const t1 = list[1] // 2
    const t2 = list[2] // 2
    console.log(head, t1, t2)
}
```

**Reply by 43081j** [reply]
`TElement` should be a generic type, and `TDocument` should be a generic type, both defined by the adapter. neither are `unknown`, they are well known types _by the adapter_. 

we want to ensure we are passing the right types to the right methods. `getAttribute(TComment)` shouldn't be possible but would be if we lazily called everything `unknown`. we'd be vastly reducing our type strength throughout the parser and covering up countless places for bugs to happen.

> But, as the types are currently broken, and we’re looking at a major release and fixing them, that’d be an acceptable trade off?

No. in google's repos alone (which use parse5), the `Node` type is used heavily for representing "some node from the AST". there's no way we should be making them change that to having to specify the union every time or create their own `Node` type around it. 100% we need to define what `Node` is for consumers, there's no way we can drop that, its a fundamental type.

in my repos, modern web, and a whole bunch of others too we all use the node type heavily.

> A tuple can be used for that right?

It wouldn't help, the logic is currently already broken. in almost every place we use the open element stack, we have assumed the current node on the stack is a `TElement`. this has never been true, because the first node is a `TDocument`. it just didn't trip us up so far because we did all this inside conditions like "if tagname == something" (which would exclude documents).

we should just have far better type inference and those conditions would automatically have narrowed the type to `TElement`.

**Reply by wooorm** [reply]
I don’t understand this conversation anymore. Whether it’s about unions or about stacks.
Everything is apparently broken, but we can’t change types (because they’re being used), instead we should remove frequently used code? (adapters, which are used *a lot*).

---

To reiterate my stance: if the current DT types are broken, and we’re creating a major to fix a bunch of stuff and add TypeScript, then stuff is going to break anyway, we might as well ship proper, working, types.

**Reply by 43081j** [reply]
no, as i've mentioned a few times i don't care if we keep the adapters or not. i care that we fix our very broken types.

yes, we should ship proper, working types. know how to write them? that's what we're currently stuck with! 😂

**Reply by wooorm** [reply]
Maybe it’s better then to have this discussion when the code is working?
It feels rather theoretical right now. Perhaps it’s better to have practical examples to look at?

**Comment by fb55** [comment]
My current thoughts — let me know if this makes sense.

Afaict what we are looking for is opaque default types for the type map. TS doesn't support that, but one way to simulate opaque types is to make the default types unique objects. `Node` etc. would still have to be defined as unions:

```ts
export interface TreeAdapterTypeMap<
    Document = { type?: 'Document' },
    DocumentFragment = { type?: 'DocumentFragment' },
    Element = { type?: 'Element' },
    CommentNode = { type?: 'CommentNode' },
    TextNode = { type?: 'TextNode' },
    Template = { type?: 'Template' },
    DocumentType = { type?: 'DocumentType' },
    ParentNode = Document | DocumentFragment | Element | Template,
    ChildNode = Element | Template | CommentNode | TextNode | DocumentType,
    Node = ParentNode | ChildNode
> {
    document: Document;
    documentFragment: DocumentFragment;
    element: Element;
    commentNode: CommentNode;
    textNode: TextNode;
    template: Template;
    documentType: DocumentType;
    parentNode: ParentNode;
    childNode: ChildNode;
    node: Node;
}
```

This seems to do the job in terms of getting good types throughout the codebase. Of course, there is a lot to fix up afterwards, but that's a separate issue.

**Reply by 43081j** [reply]
except we'd have the same problem as before:

```ts
isElement(node: Node): node is Element;
```

wouldn't work, because `Node` in your case can be anything, not necessarily something `Element` is assignable to. for that reason, `Node` can't be another type parameter, it has to be built up later on. But we can't define a type alias inside an interface, so we can't have `Node` at all. etc etc

i really think this type map is wrong, it doesn't work and only currently builds because everything narrows to `unknown`. i think we should at least throw it away and have the type params directly on the adapter interface instead.

then, though, we still have the problem above: no clean way to define what `Node` is.

you can end up with this:

```ts
type Node<T> = T extends TreeAdapter<
  infer TDocument,
  infer TElement,
  infer TComment,
  ...
> ? (TElement | TComment | ...) : never;
```

but then because `Node<T>` can be `never`, the following wouldn't work:

```ts
isElement(node: Node<this>): node is TElement; // TElement isn't assignable to never
```

we'd end up with:

```ts
isElement(node: Node<this>): node is Element<this>;
```

which would basically do the same inference to pick out `TElement` (which we frustratingly already have...)

its all a pain tbh. i don't have a solution yet. the current types are a mess, definitely broken. we're covering up a fair amount of invalid types, assignments, etc without realising it. but i haven't yet figured out how to do it properly

**Reply by 43081j** [reply]
[this](https://www.typescriptlang.org/play?#code/FASwdgLgpgTgZgQwMZQAQEEAmCAO0YA8AKgKIA2UAtlJADSpEAiA9kgK7V0MDCzlnEAHyoA3sFQTUSGFATRyVGhAAUEBAHMAXKgDOEGOHUBKbaQoCA3OMnTZ0FuwHKTDBxyVXJUmXKi9+Ss6m-pbWEupQEOgQ+iAARmzQymDMmFCmCgL0YAjU2noGYMb5sUVWAL7AwGlIZAgyUsxgeqgI2Hiw2li4+AQA5FBkffR9mKzDqH1IfJR9glbAAPSLugAWzGxkmKiwMMwwwG09sAB0EVExBglJo+MjOdR9RgtwbGBIECBNqDj1OlDETJKehMVjuLhEEJKQTKI4dGBddq9MyKCFuLI8GbQlwAN2YIG2Yi802aEEaAUgqAAvKJyq0dJiKRBPDYmi1BqiyTSRHSEAyUaEvMs1hstjsYHsDl44fgzpForFrlBlNMmfdclAniyJMKdOtNtsAO77ADWYVaSNO5wVV0Syo5GL6D01z2AlSAA) is it _actually_ strongly typed, unlike what we have now. it just means we have to repeat the node union every time, we can't put it in an alias cleanly

**Reply by fb55** [reply]
I am using a union as the default value of `Node` above, which means that `isElement` works as expected. And additional types are possible, as long as we never narrow the union. It isn't ideal, but imho a big step up from what we have now.

**Reply by 43081j** [reply]
yeah you are using a union as the default, of a type which can be anything. the default is irrelevant.

in fact the only reason this would build is because `T extends TypeMap` is telling it that it is _always_ the default types. which of course is a lie, since the real types could be of any structure. its essentially tricking the compiler into always casting to the defaults as a way to get some useful types.

for example:

```ts
type CustomTypeMap = TypeMap<
  {nodeType: 'doc'},
  {nodeType: 'frag'},
  {nodeType: 'el'},
  ...
>;

declare const adapter: Adapter<CustomTypeMap>; // ERROR, of course.
// because `{nodeType: 'doc'}` doesn't extend `{type?: 'Document'}`, our made up type
// Adapter<T extends TypeMap> === Adapter<T extends TypeMap<{type?: 'Document'}, ...>>
```

type maps don't work.

**Reply by wooorm** [reply]
Why not `T extends TypeMapWithAnys  = TypeMapWithUnknowns`?
(Or, instead of `any`s, some minimum required object form?)

**Reply by 43081j** [reply]
thats what we do now, except with `unknown`, hence the problems we're having. the same would happen with any.

```ts
interface AdapterMap {
  document: any;
  element: any;
}

interface Adapter<T extends AdapterMap> {
  getAttribute(node: T['element'], name: string): string;
}

function parse<T extends AdapterMap>(adapter: Adapter<T>): void {
    adapter.getAttribute({} as T['document'], 'x');
}
```

what its default type is doesn't matter anymore, because that isn't _the_ type. the type defines that they're all the same type (any, unknown, whatever). so you can pass any of the types as each other, i.e. no type checking happens anymore.

the only reason Felix's code worked is because he said `T extends Adapter` which will resolve to `T extends Adapter<T1, T2, ...>`. this means his constraint wasn't saying "something which extends Adapter", it was saying "something which extends the default adapter" (which none do, it was wrongly typed, i.e. a bad cast).

in his case the defaults were made up types: `Adapter<{type?: 'document'}, {type?: 'element'}, ...>`. so the code was constraining to adapters which inherit those models, _not_ any or unknown.

**Reply by wooorm** [reply]
> thats what we do now, except with `unknown`, hence the problems we're having.

So they’re not the same. **And that difference is *exactly* what I’m trying to get at.**
Using `any` for what something extends, allows adapters to type what they return.
Using `unknown` as the default means that internally, or when untyped, `unknown` is used.

**Reply by 43081j** [reply]
i am aware they're not the same, but they make no difference here. if you use `any`, you have the exact same problem: every type is assignable to each other.

nothing to stop you doing `getAttribute(node as TComment, 'foo')` if we do what you're saying. type maps don't work, because no matter what we do, we have to supply a base type of `any` or `unknown` which will result in all types being equivalent/interchangeable.

do you understand that assigning a default type doesn't solve the problem?

```ts
interface Adapter<T extends TypeMap<{
  document: {type: 'doc'};
  element: {type: 'element'};
}>> {
  getAttribute(node: T['element'], name: string): string;
  createDocument(): T['document'];
}
```

is (roughly) what felix's original suggestion actually results in. which is of course wrong, since not all adapters will use that model.

if you go any change it to `extends TypeMap<{document: any; element: any;}>`, that allows you to pass documents into methods which consume elements, etc etc.

does that make sense?

**Comment by fb55** [comment]
I have opened https://github.com/parse5/parse5-fork/pull/42 with how I think the types should work. Mostly what I posted before, with some fixes. I am struggling to find a way to sneak bad code through this anymore — @43081j please let me know if you can think of anything.
